app/entrypoint.sh

```diff
@@ -3,8 +3,8 @@
 if $APM_ENABLED; then
     echo "APM tracing enabled: $DATADOG_SERVICE_NAME"
     # start our server through ddtrace-run, which will allow us to see apm tracing in datadog
-    ddtrace-run uwsgi /config/uwsgi.yaml
+    exec ddtrace-run uwsgi /config/uwsgi.yaml
```

Comment:
How does this pattern compare with starting the process in the background and then trapping the TERM signal and waiting on the pid?

Reply:
This was the subject of a long conversation with josh, but basically: the only reason we were doing it was because we didn't understand docker well at the time.  Someone had noticed that if we had an entrypoint script and that script directly called the container's main process, then signals wouldn't be propagated to that subprocess when the container was terminated.  So, the pattern became, background that process, trap SIGTERM, wait indefinitely, then pass on SIGTERM to the subprocess when that signal was received.

Using `exec` means that dd-trace *replaces* the entrypoint script as PID 1, and so it directly receives the signal when the container is requested to terminate.  There's no need to pass the received signal to a managed subprocess because there is no longer a subprocess to manage.

Reply:
The only downside of doing it this way is that bash will wait on orphaned processes (something normally expected of PID1) whereas ddtrace may not know that it is expected to.  That said, we're already broadly not handling this case across fivestars and had decided to defer dealing with it when we met three weeks ago.

Reply:
I'd argue that the right thing to do would be to use something actually designed to be an init process in docker, like tini.  I just didn't want to introduce two new patterns at once.